include/bluetooth/services/ots_srv.h

```diff
@@ -0,0 +1,641 @@
+ *
+ *  @param feat OTS features.
+ */
+#define BT_GATT_OTS_SRV_OLCP_GET_FEAT_ORDER(feat) \
+	((feat) & BIT(BT_GATT_OTS_SRV_OLCP_FEAT_ORDER))
+
+/** @brief Get @ref BT_GATT_OTS_SRV_OLCP_GET_FEAT_NUM_REQ feature.
+ *
+ *  @param feat OTS features.
+ */
+#define BT_GATT_OTS_SRV_OLCP_GET_FEAT_NUM_REQ(feat) \
+	((feat) & BIT(BT_GATT_OTS_SRV_OLCP_FEAT_NUM_REQ))
+
+/** @brief Get @ref BT_GATT_OTS_SRV_OLCP_GET_FEAT_CLEAR feature.
+ *
+ *  @param feat OTS features.
+ */
+#define BT_GATT_OTS_SRV_OLCP_GET_FEAT_CLEAR(feat) \
+	((feat) & BIT(BT_GATT_OTS_SRV_OLCP_FEAT_CLEAR))
+
+/**@brief Features of the OTS. */
+struct bt_gatt_ots_srv_feat {
+	/* OACP Features */
+	uint32_t oacp;
+
+	/* OLCP Features */
+	uint32_t olcp;
+};
+
+/** @brief Opaque OTS instance. */
+struct bt_gatt_ots_srv;
+
+/** @brief OTS server callback structure. */
+struct bt_gatt_ots_srv_cb {
+	/** Object created callback
+	 *
+	 *  This callback is called whenever a new object is created.
+	 *  Application can reject this request by returning an error
+	 *  when it does not have necessary resources to hold this new
+	 *  object.
+	 *
+	 *  @param ots  OTS Server instance.
+	 *  @param conn The connection that is requesting object creation or
+	 *              NULL if object is created by the following function:
+	 *              bt_gatt_ots_srv_object_add().
+	 *  @param id   Object ID.
+	 *  @param init Object initialization metadata.
+	 *
+	 *  @return 0 in case of success or negative value in case of error.
+	 *  Possible return values:
+	 *  -ENOMEM if no available space for new object.
+	 */
+	int (*obj_created)(struct bt_gatt_ots_srv *ots, struct bt_conn *conn,
+			   uint64_t id,
+			   const struct bt_gatt_ots_srv_obj_metadata *init);
+
+	/** Object deleted callback
+	 *
+	 *  This callback is called whenever an object is deleted.
+	 *
+	 *  @param ots  OTS Server instance.
+	 *  @param conn The connection that deleted the object or NULL if
+	 *              this request came from the server.
+	 *  @param id   Object ID.
+	 */
+	void (*obj_deleted)(struct bt_gatt_ots_srv *ots, struct bt_conn *conn,
+			    uint64_t id);
+
+	/** Object selected callback
+	 *
+	 *  This callback is called on successful object selection.
+	 *
+	 *  @param ots  OTS Server instance.
+	 *  @param conn The connection that selected new object.
+	 *  @param id   Object ID.
+	 */
+	void (*obj_selected)(struct bt_gatt_ots_srv *ots, struct bt_conn *conn,
+			     uint64_t id);
+
+	/** Object read callback
+	 *
+	 *  This callback is called multiple times during the Object read
+	 *  operation. OTS module will keep requesting successive Object
+	 *  fragments from the application until the read operation is
+	 *  completed. The end of read operation is indicated by NULL data
+	 *  parameter.
+	 *
+	 *  @param ots    OTS Server instance.
+	 *  @param conn   The connection that read object.
+	 *  @param id     Object ID.
+	 *  @param data   In:  NULL once the read operations is completed.
```

Comment:
Now, the application fully manages buffers that are used to send Object data. OTS notifies the upper layer when the read operation is finished so that the application can do whatever it wants with the buffer (e.g. update object content, free dynamic buffer, etc.)